lang/codegen/src/generator/events.rs

```diff
@@ -0,0 +1,235 @@
+        quote! {
+            #no_cross_calling_cfg
+            #[derive(::scale::Encode, ::scale::Decode)]
+            pub enum #base_event_ident {
+                #( #event_idents(#event_idents), )*
+            }
+
+            #no_cross_calling_cfg
+            const _: () = {
+                impl ::ink_lang::BaseEvent for #storage_ident {
+                    type Type = #base_event_ident;
+                }
+            };
+
+            #(
+                #no_cross_calling_cfg
+                const _: () = {
+                    impl From<#event_idents> for #base_event_ident {
+                        fn from(event: #event_idents) -> Self {
+                            Self::#event_idents(event)
+                        }
+                    }
+                };
+            )*
+
+            const _: () = {
+                #no_cross_calling_cfg
+                impl ::ink_core::env::Topics<EnvTypes> for #base_event_ident {
+                    fn topics(&self) -> &'static [Hash] {
+                        match self {
+                            #(
+                                Self::#event_idents(event) => {
+                                    <#event_idents as ::ink_core::env::Topics<EnvTypes>>::topics(event)
+                                }
+                            )*
+                        }
+                    }
+                }
+            };
+        }
+    }
+
+    /// Generate checks to guard against too many topics in event definitions.
+    fn generate_topics_guard(&self, event: &ir::Event) -> TokenStream2 {
+        let storage_ident = self.contract.module().storage().ident();
+        let len_topics = event.fields().filter(|event| event.is_topic).count();
+        let span = event.span();
+        quote_spanned!(span=>
+            const _: () = {
+                pub enum EventTopicsOutOfBounds {}
+                pub enum EventTopicsWithinBounds {}
+                impl ::ink_lang::False for EventTopicsOutOfBounds {}
+
+                #[allow(non_camel_case_types)]
+                pub trait __ink_RenameBool {
+                    type Type;
+                }
+                impl __ink_RenameBool for [(); 1] {
+                    type Type = EventTopicsOutOfBounds;
+                }
+                impl __ink_RenameBool for [(); 0] {
+                    type Type = EventTopicsWithinBounds;
+                }
+
+                #[allow(non_upper_case_globals)]
+                const __ink_MAX_EVENT_TOPICS: usize = <
+                    <#storage_ident as ::ink_lang::ContractEnv>::Env as ::ink_core::env::EnvTypes
+                >::MAX_EVENT_TOPICS;
+
+                fn __ink_ensure_max_event_topics<T>(_: T)
+                where
+                    T: __ink_RenameBool,
+                    <T as __ink_RenameBool>::Type: ::ink_lang::False,
+                {}
+                let _ = __ink_ensure_max_event_topics::<[(); (#len_topics <= __ink_MAX_EVENT_TOPICS) as usize]>;
```

Comment:
:eyes: so what will the error look like if I exceed the max event topics?

Reply:
```rust
error[E0277]: the trait bound `erc20::_::EventTopicsWithinBounds: False` is not satisfied
  --> /home/robrob/projects/ink/examples/erc20/lib.rs:35:5
   |
35 | /     pub struct Transfer {
36 | |         #[ink(topic)]
37 | |         from: Option<AccountId>,
38 | |         #[ink(topic)]
...  |
45 | |         b: i32,
46 | |     }
   | |     ^
   | |     |
   | |_____the trait `False` is not implemented for `erc20::_::EventTopicsWithinBounds`
   |       required by this bound in `erc20::_::__ink_ensure_max_event_topics`
```

Reply:
Problem is we cannot query the `MAX_EVENT_TOPICS` value at contract compile time. So we need to force the Rust compiler into doing the check for us.

Reply:
It's probably good enough for a first iteration

Reply:
I think changing it to
```
error[E0277]: the trait bound `erc20::_::EventTopicsWithinBounds: True` is not satisfied
```
makes more sense. what do you think?

Reply:
Actually yes that does make a bit more sense! :smile: